sui_programmability/examples/games/sources/vdf_based_lottery.move

```diff
@@ -0,0 +1,156 @@
+    public entry fun complete(game: &mut Game, vdf_output: vector<u8>, vdf_proof: vector<u8>, clock: &Clock) {
+        assert!(game.status != COMPLETED, EGameAlreadyCompleted);
+        assert!(clock::timestamp_ms(clock) - game.timestamp_start >= game.submission_phase_length, ESubmissionPhaseInProgress);
+
+        // Hash combined randomness to vdf input
+        let discriminant = DISCRIMINANT_BYTES;
+        let vdf_input = sui::vdf::hash_to_input(&discriminant, &game.vdf_input_seed);
+
+        // Verify output and proof
+        assert!(sui::vdf::vdf_verify(&discriminant, &vdf_input, &vdf_output, &vdf_proof, game.iterations), EInvalidVdfOutput);
+
+        game.status = COMPLETED;
+
+        // The randomness is derived from the VDF output by passing it through sha2_256 to make it uniform.
+        let randomness = sha2_256(vdf_output);
+
+        game.winner = option::some(safe_selection(game.participants, &randomness));
+    }
+
+    /// Anyone can participate in the game and receive a ticket.
+    public entry fun participate(game: &mut Game, my_randomness: vector<u8>, clock: &Clock, ctx: &mut TxContext) {
+        assert!(game.status == IN_PROGRESS, EGameNotInProgress);
+        assert!(clock::timestamp_ms(clock) - game.timestamp_start < game.submission_phase_length, ESubmissionPhaseFinished);
+        let ticket = Ticket {
+            id: object::new(ctx),
+            game_id: object::id(game),
+            participant_index: game.participants,
+        };
+
+        // Update combined randomness
+        let pack = std::vector::empty<u8>();
+        std::vector::append(&mut pack, game.vdf_input_seed);
+        std::vector::append(&mut pack, my_randomness);
+        game.vdf_input_seed = sha2_256(pack);
+
+        game.participants = game.participants + 1;
+        transfer::public_transfer(ticket, tx_context::sender(ctx));
+    }
+
+    /// The winner can redeem its ticket.
+    public entry fun redeem(ticket: &Ticket, game: &Game, ctx: &mut TxContext) {
+        assert!(object::id(game) == ticket.game_id, EInvalidTicket);
+        assert!(option::contains(&game.winner, &ticket.participant_index), EInvalidTicket);
+
+        let winner = GameWinner {
+            id: object::new(ctx),
+            game_id: ticket.game_id,
+        };
+        transfer::public_transfer(winner, tx_context::sender(ctx));
+    }
+
+    // Note that a ticket can be deleted before the game was completed.
+    public entry fun delete_ticket(ticket: Ticket) {
+        let Ticket { id, game_id:  _, participant_index: _} = ticket;
+        object::delete(id);
+    }
+
+    public entry fun delete_game_winner(ticket: GameWinner) {
+        let GameWinner { id, game_id:  _} = ticket;
+        object::delete(id);
+    }
+
+    public fun get_ticket_game_id(ticket: &Ticket): &ID {
+        &ticket.game_id
+    }
+
+    public fun get_game_winner_game_id(ticket: &GameWinner): &ID {
+        &ticket.game_id
+    }
```

Comment:
Probably prefer not using the `get_` prefix here? Also, provide receiver syntax aliases for these?

```suggestion
    public use fun ticket_game_id as Ticket.game_id;
    public fun ticket_game_id(ticket: &Ticket): &ID {
        &ticket.game_id
    }

    public use fun game_winner_game_id as GameWinner.game_id;
    public fun game_winner_game_id(ticket: &GameWinner): &ID {
        &ticket.game_id
    }
```